libraries/Microsoft.Bot.Builder.Dialogs/MainDialog.cs

```diff
@@ -1,55 +1,97 @@
             }
 
             // Create a dialog context and try to continue running the current dialog
-            var dialogContext = await this.mainDialogSet.CreateContextAsync(turnContext).ConfigureAwait(false);
+            var dialogContext = await _mainDialogSet.CreateContextAsync(turnContext, cancellationToken).ConfigureAwait(false);
 
             // continue any running dialog
-            var result = await dialogContext.ContinueDialogAsync().ConfigureAwait(false);
+            var result = await dialogContext.ContinueDialogAsync(cancellationToken).ConfigureAwait(false);
 
             // if there isn't one running, then start main dialog as the root
             if (result.Status == DialogTurnStatus.Empty)
             {
-                result = await dialogContext.BeginDialogAsync(this.Id).ConfigureAwait(false);
+                result = await dialogContext.BeginDialogAsync(this.Id, null, cancellationToken).ConfigureAwait(false);
             }
 
             return result;
         }
+
+        protected abstract Task<DialogTurnResult> OnRunTurnAsync(DialogContext dc, CancellationToken cancellationToken = default(CancellationToken));
+
+        /// <summary>
+        /// The implementation just calls onContinueDialog()
+        /// We're overriding the components built in logic that wants to start the initial dialog.
+        /// We want the bots onRunTurn() implementation to always decide which dialog (if any) 
+        /// gets started.
+        /// </summary>
+        /// <param name="innerDC">The DialogContext scoped to this ComponnentDialog instance.</param>
+        /// <param name="options">(Optional) arguments that were passed to the dialog during `begin()` call that started the instance.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The result of running the OnRunAsync call.</returns>
+        protected override Task<DialogTurnResult> OnBeginDialogAsync(DialogContext innerDC, object options, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return OnContinueDialogAsync(innerDC, cancellationToken);
+        }
+
+        /// <summary>
+        /// The implementation calls onRunTurn()
+        /// We're overriding the components built in logic that calls innerDC.continueDialog().
+        /// We want the bots onRunTurn() implementation to decide what should happen for a turn.
+        /// </summary>
+        /// <param name="innerDC">The DialogContext scoped to this ComponnentDialog instance.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The result of running the OnRunAsync call.</returns>
+        protected override Task<DialogTurnResult> OnContinueDialogAsync(DialogContext innerDC, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return OnRunTurnAsync(innerDC, cancellationToken);
+        }
     }
```

Comment:
I don't get this. OnBegin->OnContinue ->OnRunTurn which is not defined?  WTF?

Reply:
This is a copy of Steve's. He calls Continue from Begin and then wires Continue to the abstract method which the subclass should define. I personally would prefer to tie both OnBegin and OnContinue to OnRun because I think it's cleaner. But other than that I thought sticking with Steve's was the better thing to do.